lgc/patch/PatchCheckShaderCache.cpp

```diff
@@ -102,7 +102,8 @@ bool PatchCheckShaderCache::runOnModule(Module &module) {
               vals.push_back(user);
               continue;
             }
-            return false;
+            if (getShaderStage(cast<Instruction>(user)->getFunction()) != ShaderStageFragment)
+              return false;
```

Comment:
My understanding of this is: if a Vs (for example) uses global constants, then it might put them at the end of the text section, so we wouldn't be able to merge it with a Fs to make a VsFs pipeline, because it wouldn't be able to just "fall through" into the Fs.

But is this still true? Don't we have a mechanism for forcing the Vs to put constants in a separate section and use relocs to access them?

Reply:
It is possible for LLVM to generate the data in a separate section.  There is a test that check for that https://github.com/GPUOpen-Drivers/llpc/blob/e1e567748c5b80ae06308cecfcb440f0b4663e49/llpc/test/shaderdb/relocatable_shaders/PipelineVsFs_MultipleConstData.pipe#L5.

Reply:
I know that, but I don't know how LLVM makes that decision.

I am suggesting that if a Vs has global constants, then rather than disabling partial pipeline compilation here, it would be better to do whatever it is you need to do to tell LLVM to put them in a separate section.

Reply:
> Would it be possible to have some tests written so we can see how this works.
> 
> Also, I'd like to try to understand the assumptions you are making. It seems like you are assuming that the relocations are only in the fragment shader, and that there is no prologue or epilogue for the fragment shader.

Sorry for missing the tests description.
The issue is produced when runing the following two tests with ShaderCacheMode 1 on Navi21
dEQP-VK.robustness.robustness2.bind.notemplate.r32ui.dontunroll.nonvolatile.uniform_buffer.no_fmt_qual.null_descriptor.samples_1.1d.frag (pass)
dEQP-VK.robustness.robustness2.bind.notemplate.r32ui.dontunroll.nonvolatile.uniform_buffer.no_fmt_qual.len_4.samples_1.1d.frag (fail, it has reloc section in FS but the offset and corresponding data at that offset is wrong)

Reply:
@jayfoad The condition `if (getShaderStage(cast<Instruction>(user)->getFunction()) != ShaderStageFragment)` was removed at early time as a workaround to make the above tests pass, because ELF with reloc in FS wasn't implemented in `mergeElfBinary`. Here, I restore the condition after refining `mergeElfBinary`. 
You mean we shouldn't disable partial pipeline compilation here. Could you explain more on " tell LLVM to put them in a separate section"?

Reply:
I don't know how to "tell LLVM to put them in a separate section". I was hoping someone else could explain that part, if it is possible.

Reply:
@xuechen417 The test that I linked to above tests that when relocatable shaders are enabled, the constant data is handled correctly when there is constant data in the VS.  There should be no fundamental reason to not handle a relocation for constant data.

The way this is done with relocatable shaders is to have LLVM place the data in its own section, and then the linking code makes sure the relocations are still valid.  This is the PR that enabled it: https://github.com/GPUOpen-Drivers/llpc/pull/900.

I cannot say exactly what tells the LLVM to place the data in the .rodata section.  That part was done by others.  I agree with Jay that this should be handled by placing the the constants in a different section.  This will handle even more cases.

You can try to trace what happens when --enable-relocatable-shader-elf is used and when it is not to find the difference.